bst.py

```diff
@@ -40,37 +41,165 @@ def size(self):
-            return 0
         return self._depth(self.start)
 
     def balance(self):
-        '''Returns a positive or negative value representing the bst balanced.'''
+        '''Returns a positive or negative val representing the bst balanced.'''
         if self.start is None:
             return 0
-        depth_left = self._depth(self.nodes[self.start]['left'])
-        depth_right = self._depth(self.nodes[self.start]['right'])
+        depth_left = self._depth(self.nodes[self.start].get('left'))
+        depth_right = self._depth(self.nodes[self.start].get('right'))
         return depth_left-depth_right
+
+    def get_dot(self):
+        '''Return the tree with root 'self' as a dot graph for visualization'''
+        return "digraph G{\n%s}" % ("" if self.start is None else (
+            "\t%s;\n%s\n" % (
+                self.start,
+                "\n".join(self._get_dot(self.start))
+            )
+        ))
+
+    def _get_dot(self, node):
+        """recursively prepare a dot graph entry for this node."""
+        left = self.nodes[node].get('left')
+        right = self.nodes[node].get('right')
+        if left is not None:
+            yield "\t%s -> %s;" % (node, left)
+            for i in self._get_dot(left):
+                yield i
+        elif right is not None:
+            r = random.randint(0, 1e9)
+            yield "\tnull%s [shape=point];" % r
+            yield "\t%s -> null%s;" % (node, r)
+        if right is not None:
+            yield "\t%s -> %s;" % (node, right)
+            for i in self._get_dot(right):
+                yield i
+        elif left is not None:
+            r = random.randint(0, 1e9)
+            yield "\tnull%s [shape=point];" % r
+            yield "\t%s -> null%s;" % (node, r)
+
+    def print_dot(self):
+        # To run me do this from cmd line: python bst.py > temp.png
+        import subprocess
+        dot_graph = self.get_dot()
+        t = subprocess.Popen(["dot", "-Tpng"], stdin=subprocess.PIPE)
+        t.communicate(dot_graph)
+
+    def _in_order(self, node):
+        '''Helper function for in_order method.'''
+        left = self.nodes[node].get('left')
+        right = self.nodes[node].get('right')
+
+        if left:
+            for elem in self._in_order(left):
+                yield elem
+        yield node
+        if right:
+            for elem in self._in_order(right):
+                yield elem
+
+    def in_order(self):
+        '''Return the values in the tree using in-order traversal.'''
+        for elem in self._in_order(self.start):
+            yield elem
+
+    def _pre_order(self, node):
+        '''Helper function for pre_order method.'''
+        left = self.nodes[node].get('left')
+        right = self.nodes[node].get('right')
+
+        yield node
+        if left:
+            for elem in self._pre_order(left):
+                yield elem
+        if right:
+            for elem in self._pre_order(right):
+                yield elem
```

Comment:
It strikes me that this could be cleaned up just a tiny bit:

```python
yield node
for elem in self._pre_order(self.nodes[node].get('left', [])):
    yield elem
for elem in self._pre_order(self.nodes[node].get('left', [])):
    yield elem
```

Each of your depth-first versions could benefit from this same approach.